src/components/Form/FormField.tsx

```diff
@@ -0,0 +1,207 @@
+     * @param {ReactNode|ReactNodeArray} children - The computed children for this slot.
+     */
+    renderControl: PropTypes.func,
+
+    /** The HTML input id. This will be set on the control element and will be use for linking it with the label for correct accessibility. */
+    id: PropTypes.string,
+
+    /** The HTML input name. */
+    name: PropTypes.string,
+
+    /** The HTML input type. */
+    type: PropTypes.string,
+
+    /** A label for the form field. */
+    label: customPropTypes.itemShorthand,
+
+    /**
+     * A custom render function for the label slot.
+     *
+     * @param {React.ReactType} Component - The computed component for this slot.
+     * @param {object} props - The computed props for this slot.
+     * @param {ReactNode|ReactNodeArray} children - The computed children for this slot.
+     */
+    renderLabel: PropTypes.func,
+
+    /** A field can have its label next to instead of above it. */
+    inline: PropTypes.bool,
+
+    /** A field can show that input is mandatory. */
+    required: PropTypes.bool,
+
+    /** Text message that will be displayed below the control (can be used for error, warning, success messages). */
+    message: customPropTypes.itemShorthand,
+
+    /**
+     * A custom render function for the message slot.
+     *
+     * @param {React.ReactType} Component - The computed component for this slot.
+     * @param {object} props - The computed props for this slot.
+     * @param {ReactNode|ReactNodeArray} children - The computed children for this slot.
+     */
+    renderMessage: PropTypes.func,
+
+    /** Additional CSS styles to apply to the component instance.  */
+    styles: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
+
+    /** Override for theme site variables to allow modifications of component styling via themes. */
+    variables: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
+  }
+
+  public static defaultProps = {
+    as: 'div',
+  }
+
+  public renderComponent({
+    ElementType,
+    classes,
+    accessibility,
+    variables,
+    styles,
+    rest,
+  }): React.ReactNode {
+    const {
+      children,
+      control,
+      renderControl,
+      label,
+      renderLabel,
+      id,
+      name,
+      type,
+      required,
+      message,
+      renderMessage,
+      controlType,
+    } = this.props
+
+    const labelElement = Text.create(label, {
+      defaultProps: {
+        as: 'label',
+        htmlFor: id,
+        styles: styles.label,
+      },
+      render: renderLabel,
+    })
+
+    const messageElement = Text.create(message, {
+      defaultProps: {
+        styles: styles.message,
+      },
+      render: renderMessage,
+    })
+
+    const factoryMethod = createSlotFactory(controlType, name => ({ name }))
+    const controlElement = factoryMethod(control || {}, {
+      defaultProps: { required, id, name, type, styles: styles.control },
+      render: renderControl,
+    })
+
+    const content = (
```

Comment:
now it might seem to be a matter of preference, but with more complex elements structure it will introduce unnecessary code duplication (suppose that there are three elements, instead of two `controlElement` and `labelElement` that we have now). We might consider to remove this source for code duplication in the following way:

```jsx
<>
  {isControlElementFirst && controlElement}
  {labelElement}
  {!isControlElementFirst && controlElement}

  {messageElement}
</>
```

Reply:
Agreed looks much more cleaner.